### PR TITLE
Added Group `pin`

### PR DESCRIPTION
**Pull request by RichardLitt**
I am not sure about ls count. I do not think that ls --quiet is an argument: the go code is a flag for Fprintf or not, which makes me think it does not affect the API returns. But I cannot tell, because in Wireshark it is all sent as flags.

### CODE REVIEW

File: apiary.apib [hunk 1]
```diff
@@ -1682,12 +1682,476 @@ root will not be listable, as it is virtual. Access known paths directly.
...
 
 # Group pin
+Pin (and unpin) objects to local storage
 
-## add
+This command can't be called directly.
 
-## ls
+## add [GET /pin/add{?arg}{&recursive}]
+Pins objects to local storage
```
Comment (dignifiedquire) [inline apiary.apib]: Missing `.`

File: apiary.apib [hunk 2]
```diff
@@ -1682,12 +1682,476 @@ root will not be listable, as it is virtual. Access known paths directly.
...
+
+Retrieves the object named by <ipfs-path> and stores it locally
+on disk.
+
++ Parameters
+    + arg (string, required) - Path to object or objects to be pinned
+    + recursive (boolean, optional) - Recursively pin the object linked to by the specified object or objects
```
Comment (dignifiedquire) [inline apiary.apib]: Default value is `true` in 0.4

File: apiary.apib [hunk 3]
```diff
@@ -1682,12 +1682,476 @@ root will not be listable, as it is virtual. Access known paths directly.
...
+    + Body
+
+        ```
+        {
+          "Message": "pin: invalid ipfs ref path",
+          "Code": 0
+        }
+        ```
+
++ Request With Argument
```
Comment (dignifiedquire) [inline apiary.apib]: Would be good to have a recursive example

File: apiary.apib [hunk 4]
```diff
@@ -1682,12 +1682,476 @@ root will not be listable, as it is virtual. Access known paths directly.
...
+
++ Parameters
+    + type (enum[string], optional) - The type of pinned keys to list.
+        Default: `recursive`
+        Members:
+            + `all`
+            + `direct`
+            + `indirect`
+            + `recursive`
+    + count (boolean, optional) - Show refcount when listing indirect pins
```
Comment (dignifiedquire) [inline apiary.apib]: Default value?

File: apiary.apib [hunk 5]
```diff
@@ -1682,12 +1682,476 @@ root will not be listable, as it is virtual. Access known paths directly.
...
+    + quiet (boolean, optional) - Write just hashes of objects
```
Comment (dignifiedquire) [inline apiary.apib]: Default value?

File: apiary.apib (hunk 2 above)
Comment (RichardLitt) [inline apiary.apib]: Not seeing this in the request on wireshark after calling `curl -i -X POST "http://localhost:5001/api/v0/pin/add?arg=QmW2WQi7j6c7UgJTarActp7tDNikE4B2qXtFCfLPdsgaTQ"`

File: apiary.apib (hunk 4 above)
Comment (RichardLitt) [inline apiary.apib]: Unable to get working, true or false.

File: apiary.apib (hunk 5 above)
Comment (RichardLitt) [inline apiary.apib]: Also unable to get working on the API. @whyrusleeping Why is that?

File: apiary.apib (hunk 2 above)
Comment (dignifiedquire) [inline apiary.apib]: It's set inside go-ipfs so you won't see it in the request

Comment (RichardLitt) [inline apiary.apib]: In that case, oughtn't we to delete it from here?

Comment (dignifiedquire) [inline apiary.apib]: Why delete, you can still run with `recursive=false`

Comment (dignifiedquire) [inline apiary.apib]: Just note that the default is `true`

Comment (RichardLitt) [inline apiary.apib]: Because it doesn't make a difference whether we run it with or without, as far as I can tell.

Comment (dignifiedquire) [inline apiary.apib]: That would be a bug, you should be able to pin a top an object with links, without pinning the links if you want to. Why else would this option even exist?

**Comment by RichardLitt** [comment]
@whyrusleeping Note @dignifiedquire's comments in the outdated diffs. Some of the flags don't seem to work on the API.

**Comment by RichardLitt** [comment]
Before merging, this needs to be updated to reflect changes in https://github.com/ipfs/go-ipfs/pull/2477/files.

**Comment by daviddias** [comment]
@RichardLitt https://github.com/ipfs/go-ipfs/pull/2477 was merged. Can we merge this one now?

**Comment by RichardLitt** [comment]
Merged!

**Comment by daviddias** [comment]
wooo :)

**Pull request merged**